c7n/resources/opsworks.py

```diff
@@ -31,6 +57,101 @@ class resource_type(object):
         dimension = "StackId"
 
 
+@OpsworkStack.action_registry.register('delete')
+class DeleteStack(BaseAction, StateTransitionFilter):
+    """Action to delete Opswork Stack
+
+    It is recommended to use a filter to avoid unwanted deletion of stacks
+
+    :example:
+
+        .. code-block: yaml
+
+            policies:
+              - name: opswork-delete
+                resource: opswork-stack
+                actions:
+                  - delete
+    """
+
+    valid_origin_states = ('terminating', 'stopping', 'shutting_down', 'terminated', 'stopped')
+
+    schema = type_schema('delete')
+    permissions = ("opsworks:DescribeApps", "opsworks:DescribeLayers",
+        "opsworks:DescribeInstances", "opsworks:DeleteStack",
+        "opsworks:DeleteApp", "opsworks:DeleteLayer",
+        "opsworks:DeleteInstance")
+
+    def process(self, stacks):
+        with self.executor_factory(max_workers=2) as w:
+            list(w.map(self.process_stacks, stacks))
+
+    def process_stacks(self, stack):
+        client = local_session(
+            self.manager.session_factory).client('opsworks')
+        try:
+            stack_id = stack['StackId']
+            for app in client.describe_apps(StackId=stack_id)['Apps']:
+                client.delete_app(AppId=app['AppId'])
+            instances = client.describe_instances(StackId=stack_id)['Instances']
+            orig_length = len(instances)
+            instances = self.filter_instance_state(instances)
+            if(len(instances) != orig_length):
+                self.log.exception("All instances must be stopped before deletion.")
+                return
+            for instance in instances:
+                instance_id = instance['InstanceId']
+                retryable = ('ValidationException'),
+                retry = utils.get_retry(retryable, max_attempts=8)
```

Comment:
whats the validation exception about? please put a comment in here regarding it

Reply:
This validation exception occurs if the instance is still stopping when delete is called. I included this retry functionality as stopping a stack often takes a decent amount of time, so the code allows "stopping"/"shutting-down" as one of the valid origin states. "Validation Exception" is returned if delete is called, but the instance is still stopping. If we want the function to only permit deletion when everything is completely stopped at function call (instead of allowing a few retries to permit instances to finish stopping), I can remove this try-except statement.